enhancements/cluster-logging/forwarding-json-structured-logs.md

```diff
@@ -0,0 +1,385 @@
+```yaml
+outputs:
+- name: jsonOut
+  ...
+  structured:
+    discard: none
+pipelines:
+- inputRefs: [ application ]
+  outputRefs: [ jsonOut ]
+```
+
+#### I want to  forward structured and unstructured logs from the same input to different outputs
+
+```yaml
+outputs:
+- name: structuredOut
+  ...
+  structured:
+    discard: unstructured
+- name: unstructuredOut
+  ...
+  structured:
+    discard: structured
+pipelines:
+- inputRefs: [ application ]
+  outputRefs: [ structuredOut, unstructuredOut ]
+```
+
+### Proposal details
+
+#### Data model changes
+
+Two new top-level fields added to the internal [data model][model]:
+* `structured_message` (type object): Original structured JSON log message.
+* `structured_format` (type string): Format name from `input.structured.format`.
+
+Note: `structured_format` is an arbitrary name assigned by the user, it has no meaning to the forwarder.
+
+The following rules apply to forwarded records:
+* At least one of `message` or `structured_message` MUST be present. They MAY both be present.
+* `message` and `structured_message` MAY both be present, if so `message` MUST be the JSON quoted-string form of the value of `structured_message`.
+* `structured_format` MAY be present if `structured_message` is present, and MUST NOT be present if `structured_message` is missing.
+
+Interpretation of the record:
+* `structured_format` is present: it indicates the format of `structured_message`
+* `structured_format` is missing:
+  * `structured_message` is present it: it is JSON, but of an unknown format.
+  * `structured_message` is missing: `message` is the original message of unknown encoding.
+* If both `structured_message` and `message` are present they are redundant copies of the same data. May be useful in upgrade transitions.
+
+#### ClusterLogForwarder configuration
+
+New field `input.stuctured` with sub-field
+* `format`: Name to forward as value of `structured_format`. If missing, no `structured_format` field is forwarded.
+
+New field `output.structured` with sub-fields:
+* `discard` (enum: Structured, Unstructured) discard either structured or unstructured messages. If missing do not discard anything.
+* `unstructured`: enum indicating treatment of unstructured `message` field.
+  - Only: Send *only* `message`, without `structured_message`, even if parsed JSON is available.
+  - Copy: Send `message` and `structured_message` (if available). For upgrade transitions.
+  - (missing): If parsed JSON is available send only `structured_message` , else only `message`
+
+### Implementation Details
+
+#### Enabling, optimizing JSON parsing
+
+In the current `fluentd` collector, JSON parsing must be enabled at the source.
+It is enabled if either of the following is true:
+* The corresponding `input` has a `structured` field.
+* Any connected `output` has a `structured` field.
+
+Note: if the `input.structured` field is present, then all outputs may receive structured JSON.
```

Comment:
is there any reason we should hide JSON behind "structed" if it must be JSON?  any anticipation that i could ever be something else w may wish to consume?

Reply:
Here are some possibilities, I'm not sure if we actually will ever see demand for them.

An rsyslog structured message would put metadata in USERDATA name/value sections, and put the unmodified original log text in the MSG field for the original message - that would be more in keeping with rsyslogs design. Currently we drop the JSON message into the MSG field.

An AMQP structured message would use AMQP application properties.
An MQTT structured message would use MQTT user properties.
An XML structured message would ... no, please, don't drag me back there!!!!
etc... 

With all of the above transports we can also send a structured JSON payload and ignore their native ways of adding metadata, so I may be over-anticipating. It comes down to whether end users would prefer everything in JSON regardless of transport, or are committed to some transport and would prefer to use their existing native libraries to pick apart messages rather than add JSON parsers.

And there's the fact that, though currently fashionable,  JSON itself sucks in multiple ways, so we'll soon have The Next Great Half-Baked Universal Encoding thrust upon us.

Reply:
See latest.